### PR TITLE
[feat]mmprocessor: support minicpm prep pixel values input

### PR DESCRIPTION
**Pull request by cjhgo**
## Motivation
follow the great work of 
https://github.com/sgl-project/sglang/issues/6483
https://github.com/sgl-project/sglang/pull/6543
in this pr, i try to adjust minicpm mmprocessor to support preprocessed pixels values input.

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by JustinTong0323** [comment]
Great work! Thanks for the contribution! May you first resolve the conflicts?

**Comment by cjhgo** [comment]
> Great work! Thanks for the contribution! May you first resolve the conflicts?

sure, i will resolve the conflicts and test compatibility again.

**Comment by cjhgo** [comment]
## Question: How to Handle Multi-Image Input for Preprocessed Data?

hi there! I've tested the latest code and verified that single-image input works correctly. 
This improvement is very helpful for improving throughput! 

Now I have a question about how to properly handle multi-image inputs with preprocessed data.

### Current Understanding

I'm looking at the `load_mm_data` method and trying to understand how it handles multi-image scenarios with preprocessed inputs. The method supports both raw and preprocessed inputs through the `_load_single_item` method at line ~198:

```python
def _load_single_item(data, is_video, is_audio, frame_count_limit=None, discard_alpha_channel=True):
    if isinstance(data, dict):
        return data  # Direct return for preprocessed data
```

### My Question

For multi-image VLM scenarios, I notice there seems to be a format difference:

**Raw multi-image input:**
- Format: `image_data: [img0, img1]`
- Each image is processed independently
- Result: `images = [processed_img0, processed_img1]`  
- Token generation: 2 tokens (one per image)

**Preprocessed multi-image input:**
- Format: `dict(pixel_values=...)` where `pixel_values` contains merged results of multiple image slices/patches
- For MiniCPM VLM preprocessing, **multiple images are combined into a single dict**
- Result: `images = [dict(pixel_values=...)]`
- Token generation: 1 token (but contains multiple images)

### What I'm Wondering


How should I properly handle multi-image inputs when using preprocessed data? Should the preprocessed format include information about the number of images, or is there a different approach you'd recommend?

I want to make sure I'm handling this correctly to maintain consistency between raw and preprocessed multi-image inputs.

Thanks for any guidance!

---

**Comment by JustinTong0323** [comment]
Hey @cjhgo , this is a really relevant question. Thanks a lot for pointing it out to us. Right now, the instructions about preprocessed data are kind of unclear. To make it clear, we figure that each input data item should be a single thing. Like for images, the input image_data should be a list of individual ImageDataItem. Each of these could be a dictionary, holding all the necessary info for one image, like pixel values and such.

**Comment by cjhgo** [comment]
## Update: Multi-Image Input Compatibility Implementation

hey @JustinTong0323 thanks for the golden suggestion!
I've implemented the multi-image input compatibility for preprocessed data as discussed earlier. 
Here are the key changes:

### Problem Solved

Previously, there was an inconsistency between raw and preprocessed multi-image inputs:
- **Raw multi-image**: `image_data: [img0, img1]` → generates 2 tokens  
- **Preprocessed multi-image**: `image_data: [dict(pixel_values=...)]` → generated only 1 token (but contained multiple images)

### Solution Implemented

**1. Modified `_pixel_values_image_data` method in MiniCPM test class:**
```python
def _pixel_values_image_data(self, processor_output):
    image_slice_patch_size = len(processor_output.pixel_values[0])
    pre_dicts = []
    for i in range(image_slice_patch_size):
        pre_dicts.append(dict(
            modality="IMAGE",
            pixel_values=processor_output.pixel_values[0][i],
            tgt_size=processor_output.tgt_sizes[0][i],
        ))
    return pre_dicts
```

Instead of returning a single merged dict, we now decompose preprocessed multi-image data into individual dict items, with each dict containing `pixel_values` and `tgt_size` for one image slice/patch.

**2. Enhanced `process_mm_data_async` method:**
- Added proper handling for both merged and individual preprocessed data formats
- Implemented data flattening logic for consistent processing
- Maintained backward compatibility with existing single-image preprocessed inputs

**3. Added comprehensive testing:**
- `test_understands_pixel_values_with_imgs`: Tests multi-image input with two different images
- Verifies that the model correctly processes multiple images in preprocessed format


### Verification

I've tested this implementation with:
- Single image preprocessed inputs ✅ (confirmed working, great throughput improvement!)
- Multi-image preprocessed inputs ✅ (now properly supported)


This solution ensures that `load_mm_data` correctly handles multi-image compatibility for preprocessed inputs while preserving all the performance benefits of the original implementation.

### Questions for Reviewers

**1. Implementation Feasibility:** 
Could you please help evaluate whether this current implementation approach is feasible and aligns with the overall architecture design? I want to make sure this solution is robust and follows best practices.

**2. Test Coverage:** 
I've added multi-image validation methods to the test file, including `test_understands_pixel_values_with_imgs` which tests scenarios with multiple different images. However, this has made the test file considerably longer. 

**Should I keep these extended test cases, or would you prefer them to be:**
- Moved to a separate test file?
- Simplified to cover only essential scenarios?
- Kept as is for comprehensive coverage?

I'd appreciate your feedback on both the technical approach and the testing strategy!

File: test/srt/test_vlm_input_format.py [hunk 1]
```diff
@@ -215,6 +215,77 @@ def _pixel_values_image_data(self, processor_output):
...
+        req = self.get_completion_request()
+        processor_output = self.get_processor_output(req=req)
+        output = await self.engine.async_generate(
+            input_ids=processor_output["input_ids"][0].detach().cpu().tolist(),
+            image_data=self._pixel_values_image_data(processor_output),
+            sampling_params=dict(temperature=0.0),
+        )
+        self.verify_response(output)
+
+    async def test_understands_pixel_values_with_imgs(self):
```
Comment (JustinTong0323) [inline test/srt/test_vlm_input_format.py:252]: This test is good! Maybe we could move to VLMInputTestBase.

File: test/srt/test_vlm_input_format.py [hunk 2]
```diff
@@ -215,6 +215,77 @@ def _pixel_values_image_data(self, processor_output):
...
+
+    async def test_understands_pixel_values(self):
+        req = self.get_completion_request()
+        processor_output = self.get_processor_output(req=req)
+        output = await self.engine.async_generate(
+            input_ids=processor_output["input_ids"][0].detach().cpu().tolist(),
+            image_data=self._pixel_values_image_data(processor_output),
+            sampling_params=dict(temperature=0.0),
+        )
+        self.verify_response(output)
```
Comment (JustinTong0323) [inline test/srt/test_vlm_input_format.py:250]: This test seems the same as that in base?

Comment (cjhgo) [reply test/srt/test_vlm_input_format.py:250]: > This test seems the same as that in base?

some tiny difference.
with `input_ids=processor_output["input_ids"][0].detach().cpu().tolist()`
TokenizerManager.generate_request will construct a single GenerateReqInput.

when image_data is single, it is okay to write `image_data=[self._pixel_values_image_data(processor_output)]` 
which will be `image_data=[dict]`.
when image_data is multi image for one prompt, i'd like to  pass `image_data=[dict,dict,dict...]`.
then `image_data=self._pixel_values_image_data(processor_output)` at base will not fit,
it will resolve to `image_data=[[dict,dict,dict...]]`.
so i have to remove the outside `[]` in the custom sub method.


on the other hand，
if we provide `input_ids=[processor_output["input_ids"][0].detach().cpu().tolist()]` like `test_understands_pixel_values_with_imgs`,
then TokenizerManager.generate_request will construct a GenerateReqInput whose is_single = False,
it is okay to write `image_data=[self._pixel_values_image_data(processor_output)]` .

input type is so flexible that we have to handle the subtle `[]`.

Comment (JustinTong0323) [reply test/srt/test_vlm_input_format.py:250]: I think for a single request with multiple images, `image_data=[[dict,dict,dict...]]` is acceptable.

Comment (JustinTong0323) [reply test/srt/test_vlm_input_format.py:250]: > input type is so flexible that we have to handle the subtle [].

Yes, it confused me a lot at beginning. Maybe we could try to figure out a better way, or fix a pattern for the input format, as it's only for offline inference.

Comment (cjhgo) [reply test/srt/test_vlm_input_format.py:250]: > I think for a single request with multiple images, `image_data=[[dict,dict,dict...]]` is acceptable.

Also reasonable.
let me figure out how to handle this case 
so that no need to duplicate most base test code ~
